godot-core/src/builtin/arrays.rs

```diff
@@ -6,12 +6,30 @@
 
 use godot_ffi as sys;
 
-use crate::builtin::{inner, FromVariant, Variant};
+use crate::builtin::{inner, FromVariant, ToVariant, Variant, VariantConversionError};
+use std::fmt;
 use std::marker::PhantomData;
 use sys::types::*;
 use sys::{ffi_methods, interface_fn, GodotFfi};
 
-impl_builtin_stub!(Array, OpaqueArray);
+/// Godot's `Array` type.
+///
+/// This is a variant array, meaning it contains `Variant`s which may be of different types even
+/// within the same array.
+///
+/// Unlike GDScript, all indices and sizes are unsigned, so negative indices are not supported.
+///
+/// # Safety
+///
+/// Usage is safe if the `Array` is used on a single thread only. Concurrent reads on different
+/// threads are also safe, but any writes must be externally synchronized. The Rust compiler will
+/// enforce this as long as you use only Rust threads, but it cannot protect against concurrent
+/// modification on other threads (e.g. created through GDScript).
```

Comment:
One thing worth mentioning is that each `Array` instance lives independently and has copy semantics (in fact CoW, but from the user perspective, copying an array creates an independent on).

Can you imagine any scenario where this leads to safety issues? E.g. the old instance is iterated over, but a copy (yet-to-be-CoW-written) is now modified.

Reply:
> Isn't that always true, for any kind of data structure? If all the accesses are read-only, threading is never a problem, many functional languages are built on this principle.

It's true that concurrent reads are normally safe, but concurrent writes can also be made safe internally (using mutexes or whatever). The main point is that we don't do that.

> Is the safety-related information specific to Array or should we maybe move it one level up, to the `builtin` module documentation?

I just wrote it here for now because I felt bad about ignoring GDScript threads altogether, whereas gdnative's `VariantArray` [handles them more explicitly](https://docs.rs/gdnative/latest/gdnative/core_types/struct.VariantArray.html#safety). But I didn't feel qualified to make such safety claims for other parts of the gdextension :) If it needs to go somewhere, I'd expect it _even_ higher up, at the top level.

> One thing worth mentioning is that each `Array` instance lives independently and has copy semantics (in fact CoW, but from the user perspective, copying an array creates an independent on).

`Array` is a strange beast. It's [refcounted](https://github.com/godotengine/godot/blob/e93266b9ff359c98e9f8e2a550e16ad77490fc4d/core/variant/array.cpp#L46), so writes _are_ reflected in all copies, but the refcounted thing is `Vector<Variant>` which itself is a [refcounted copy-on-write array](https://github.com/godotengine/godot/blob/e93266b9ff359c98e9f8e2a550e16ad77490fc4d/core/templates/vector.h#L66).

So the question is: can it ever happen that the inner `Vector<T>` is shared between two seemingly independent instances of `Array`? I don't think so, but I wouldn't bet on it.

And the other question is: does it matter? Surely `CowData` would create a copy _before_ applying any mutations, right?

Reply:
I don't see through the entire `Array` implementation in Godot... Would need to spend more time on it.

What I noticed:
* Two instances share state (the ref-counter), when one instance is copied from the other. But it looks like Godot uses the `SafeRefCount` class, which uses `std::atomic` internally. So concurrent modifications of the ref-counter should not cause race conditions or UB.
* `Array` has different methods that look like copying:
   * `_assign(const Array&) -> bool`
   * `typed_assign(const Array&) -> bool)
   * `duplicate() -> Array`
   * `Array(const Array &p_from)`
   I'm not sure which one is used by the FFI copy constructor method, would need to run the debugger.

I think we're good for now. Let's get this initial implementation running.